### ISSUE TITLE
Web client hanging forever

### ISSUE THREAD

**Issue by yadid** [mher/flower#639]
Im trying to run flower on my local celery deployment. When I try to connect via web browser to localhost:5555 or via curl , it just hangs forever (waiting for localhost). I can telnet to port 5555, but other than that no response. Below is some debug info:

celery -A myproject flower --broker=redis://localhost:6379/0 --broker_api=redis://localhost:6379/0 --debug --inspect_timeout=10
[I 161120 20:08:01 command:136] Visit me at http://localhost:5555
[I 161120 20:08:01 command:141] Broker: redis://localhost:6379/0
[I 161120 20:08:01 command:144] Registered tasks:
    [u'app.tasks.generate_scheduled',
     u'celery.accumulate',
     u'celery.backend_cleanup',
     u'celery.chain',
     u'celery.chord',
     u'celery.chord_unlock',
     u'celery.chunks',
     u'celery.group',
     u'celery.map',
     u'celery.starmap',
     u'app.tasks.generate_result',
     'normalize_time_task',
     u'myplatform.celery.debug_task',
     u'tasks.bake_block',
     'tasks.baker',
     'tasks.mailman']
[D 161120 20:08:01 command:146] Settings: {'cookie_secret': 'pnKuJkmuRPW43p1N4KYmvRAZjwFJtE+dvDWUNGrHTl8=',
     'debug': True,
     'login_url': '/login',
     'static_path': '/usr/local/lib/python2.7/site-packages/flower-0.9.1-py2.7.egg/flower/static',
     'static_url_prefix': '/static/',
     'template_path': '/usr/local/lib/python2.7/site-packages/flower-0.9.1-py2.7.egg/flower/templates'}
[D 161120 20:08:01 control:29] Updating all worker's cache...
[I 161120 20:08:01 mixins:224] Connected to redis://localhost:6379/0

**Comment by toorsukhmeet** [comment]
Having the same issue with RabbitMQ broker. Workers work as expected but flower (0.9.1) hangs.

**Comment by thenewguy** [comment]
I have recently started observing this as well and have not been able to track down what started the issue

**Comment by thenewguy** [comment]
Master did not fix the issue

**Comment by thenewguy** [comment]
My problem was ssl connection problems. Unrelated to flower... was a certificates permissions error on the filesystem

**Comment by enriquefernandez** [comment]
I'm having the same issue with Redis in OSX El Capitan.

Going to: http://localhost:5555/ hangs. It doesn't display anything.
In the terminal, flower doesn't complain about anything. It seems to start normally and it even displays the correct names of the registered tasks.

**Comment by philejmath** [comment]
I've having the same issue. Flower hanging indefinitely. Seems to start normally and displays all registered task.

**Comment by Jahangir-Sh** [comment]
Same issue, hangs forever. There is nothing descriptive logs on debug mode `--debug`.

**Comment by kellanburket** [comment]
I ran into this issue while using Redis Sentinel to connect to Flower. One of the Sentinel IPs wasn't resolving, but Flower didn't display an error message and would hang when I tried to connect to it. It's extremely frustrating that there was no error warning when my broker couldn't be found. It's also unclear to me, why this failure resulted in an inability to set up the web client--at the very least Flower should display some kind of error message in this case.

**Comment by MegacoderKim** [comment]
Every time i have experienced this is usually a celery config issue with the log suppressed by flower.
try invoking the worker alone and see if the logs give any leads
`celery -A app worker --loglevel=inf0`

**Comment by ocervell** [comment]
Same issue here ... Happened after I switched to eventlet or gevent to run my tasks.

**Comment by senyuuri** [comment]
I'm having the same issue here. 

If you are using WSL/Bash on Windows, this might be a result a of [a known compatibility issue](https://github.com/celery/py-amqp/pull/167) between `amqp 2.1.1 - 2.2.2` and WSL. (It's fixed in `amqp 2.2.3`, which has not been released at this moment). 

In my case, I was using `celery==4.1.0` and `amqp==2.1.1`. Everything was working fine. However, if I install `flower` through
`pip install flower`, run `celery -A myprojectname flower` and open `http:\\localhost:5555` in the browser, the page just hangs.

I quickly ran a `pip freeze` and noticed my amqp version was replace with the current latest 2.2.2. Reinstalling an older version of amqp simply solves the issue. `http:\\localhost:5555` now correctly shows the flower dashboard.

Hope this can help.

**Comment by djlambert** [comment]
Just ran into this issue. I added the following lines to the beginning of my Flask app to fix an issue with websockets.
```python
import eventlet
eventlet.monkey_patch()
```
Commenting this out resolved the issue with Flower, so as @ocervell mentioned eventlet is involved somehome. Will need to find a workaround.

**Comment by gwsampso** [comment]
Having similar problems with 
apache-airflow==1.9.0
amqp==1.4.9
celery==3.1.25
flower==0.9.2

**Comment by shashank-iitj** [comment]
I encountered similar problem as well with:
amqp==2.2.2
celery==4.1.0
flower==0.9.2
kombu==4.1.0

**Comment by mberacochea** [comment]
Same issue:

flower==0.9.2
celery==4.1.1
amqp==2.3.1
kombu==4.2.0

**Comment by erindy** [comment]
Having the same problem:
kombu (4.1.0)
Django (2.0.4)
celery (4.1.0)
amqp (2.2.2)

**Comment by djlambert** [comment]
I switched from ```eventlet``` to ```gevent``` (dependency of another package) and updated the ```bin/celery``` script to import gevent and monkey patch before all the other imports. I can now run flower with ```celery flower --app myapp.app --pool gevent``` without issue. Also my flower==1.0.0 and celery==4.2.0 now.

**Comment by thestevenbell** [comment]
Flower was hanging for me, added gevent via pip3 and started with `celery flower -A src.celery --address=127.0.0.1 --port=5555 --logLevel=debug --pool gevent`

**Comment by tomli1981** [comment]
Any solution for this issue ?

**Comment by zetaab** [comment]
with --pool gevent I cannot see any workers online

**Comment by sivang** [comment]
I'm getting this when trying to click one the workers and inspect it's configs and details , any idea how to solve this or how to make flower be able to resolve DNS for the worker if that is indeed the problem?

**Comment by maciejzukowski** [comment]
I performed a 'FLUSHALL' on redis via the redis-cli and then restarted flower and it started to work.

**Comment by maribedran** [comment]
Having the same problem with docker. Running directly on the machine flower works normally, but inside the container it hangs forever.

```
amqp==2.3.2
celery==4.1.1
flower==0.9.2
kombu==4.2.1
tornado==5.1.1
```

**Comment by pdoshi-ns** [comment]
I believe you have to start the redis server first. If there is no active redis server, flower does not give you an error but it will just keep looking for one and trying to connect to it resulting in it hanging. So to start a simple redis server just type redis-server in your terminal. After that run flower.

**Comment by chenliang0571** [comment]
I have the same issue, in my case it can be resolved by adding **--config** argument.

`flower -b 'sentinel://192.168.1.1:26379;sentinel://192.168.1.2:26379' --address=0.0.0.0 --port=5555 --config=airflow.config_templates.default_celery`

```
$ flower --version
0.9.2
4.1.1 (latentcall)
flower   -> flower:0.9.2 tornado:5.1.1 babel:2.6.0
software -> celery:4.1.1 (latentcall) kombu:4.3.0 py:3.6.5
            billiard:3.5.0.5 py-amqp:2.4.1
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:amqp results:disabled
```
checkout this jira for detail:
https://issues.apache.org/jira/browse/AIRFLOW-5272

**Comment by gainskills** [comment]
For my case, I examed the connection between flower and rabbitmq / redis, got the issue resolved after the connectivity works well.

**Comment by sizzlorox** [comment]
I'm not sure if this is the same issue or not but I've been having problems where my celery was getting heartbeat timeout issues with rabbitmq as a broker. I followed a few workarounds where people disabled the hearbeat for rabbitmq and celery. This fixed the timeout problem but ended up causing this bug with flower where about 5 seconds after booting up flower, the webclient would hang forever. After adding the heartbeat back, my flowers webclient has gone back to normal.

**Comment by bartolootrit** [comment]
I have flower-0.9.3 and tornado-5.1.1. After downgrading Tornado to v4.5.2 Flower works correctly (see https://github.com/mher/flower/issues/788#issuecomment-373622016)

**Comment by mher** [comment]
Can someone try https://github.com/mher/flower/compare/issue639 fix?

**Comment by sillycube** [comment]
Any fix in 2020? Mine is working well in dev server, but it won't work in prod server. 
docker-compose.override.yml
`flower:
    image: mher/flower
    environment:
      - CELERY_BROKER_URL=redis://redis:6379
      - FLOWER_PORT=5555
      - FLOWER_BASIC_AUTH=${FLOWER_USER}:${FLOWER_PASSWORD}
    ports:
      - "5555:5555"`

**Comment by ckcollab** [comment]
@sillycube any output from flower? Everything appears to be running fine on prod for us, but we're also experiencing no response/hanging..

`docker-compose.yml`:

```
  flower:
    image: mher/flower
    env_file: .env
    environment:
      - CELERY_BROKER_URL=pyamqp://${RABBITMQ_DEFAULT_USER}:${RABBITMQ_DEFAULT_PASS}@rabbit:${RABBITMQ_PORT}//
    ports:
      - ${FLOWER_PUBLIC_PORT:-5555}:5555
```

**Comment by ckcollab** [comment]
I _believe_ the problem has been narrowed down to `RABBITMQ_PORT` running on a non-default port causing problems? As far as I know `CELERY_BROKER_URL` is the only variable that needs to account for the different port, am I missing another change that's needed?

**Comment by sillycube** [comment]
Hi @ckcollab,

I tried to change the ports from 5555:5555 to 5556:5555 and I can access flower with the new link

**Comment by ckcollab** [comment]
The problem for us ended up being the hostname wasn't resolving over docker properly, on the different port. On port `5672` `rabbit` docker container hostname worked fine, on `9002` it didn't work.. we had to use `docker.for.mac.localhost` or the IP of the actual server with port `9002`... strangely..

**Comment by timothywangdev** [comment]
This is due to flower not being able to connect to the broker.
If you are docker-compose file, make sure that your broker ip address is the broker container name.

```
version: '3'
services:
    rabbitmq:
        restart: always
        image: rabbitmq:management
        volumes:
            - "./data/rabbitmq:/var/lib/rabbitmq"
        ports:
            - "5672:5672"
            - "15672:15672"
    flower:  
        image: mher/flower
        command: ["flower", "--broker=amqp://rabbitmq:5672"]  
        ports:  
            - 5555:5555  
        depends_on:
            - rabbitmq
```

Here, "--broker=amqp://rabbitmq:5672" is used to connect to the "rabbitmq" container
Also, turn on the debug option in flower might be helpful.

**Comment by cmoiccool** [comment]
Hi,

In my case it was able to connect with the broker, so this was not the issue. I tried the [version from branch issue639](https://github.com/mher/flower/issues/639#issuecomment-569631223) but it did not solve the problem.
The issue i was having was some incompatibility with the Elastic APM client that was installed in my Django App. Removing `'elasticapm.contrib.django',` from `INSTALLED_APPS` solved it.

So for the ones that still have not found how to solve this, this might be one of the possible cause to investigate.

**Comment by NBardelot** [comment]
I have the same issue with Airflow configured with CeleryExecutor and Flower. 

We run everything using the stable Airflow Helm chart, and in the pod that runs Flower I can confirm that I can access my Redis broker using OpenSSL `s_client` (as our Redis is a TLS-encrypted endpoint):

`s_client -connect <redis host>:6379`

I provide the remote certificate as a PEM in `/etc/ssl/certs`. And the encrypted connexion is successfuly created, and I can use commands like AUTH and INFO to talk to my Redis server.

That said, the Flower server itself does not seems to work. It hangs after some initial logs. It is not reachable through port 5555 (even if the socket *is* listening). 

I've published a couple of PRs for [puckel's Docker image of Airflow](https://github.com/puckel/docker-airflow) in order to be able to use manual configuration for Airflow using Docker: 

https://github.com/puckel/docker-airflow/pull/495
and
https://github.com/puckel/docker-airflow/pull/503

And I use the following configuration for every Airflow component *including* Flower:

`AIRFLOW__CELERY__BROKER_URL=redis://:<basic auth>@<redis host>:6379/<queue number>`


While it works for Airflow itself (i.e. the scheduler), it does not work for Flower. In my opinion the issue is that Flower does not decrypt the TLS connection the same way as the other Airflow components.

As stated in the Celery documentation: 

https://docs.celeryproject.org/en/stable/userguide/configuration.html#id22

> This backend requires the result_backend setting to be set to a Redis or Redis over TLS URL:
> ...
> Use the rediss:// protocol to connect to redis over TLS:

`result_backend = 'rediss://:[email protected]:port/db?ssl_cert_reqs=required'`

> Note that the ssl_cert_reqs string should be one of required, optional, or none (though, for backwards compatibility, the string may also be one of CERT_REQUIRED, CERT_OPTIONAL, CERT_NONE).

But the issue here is that Airflow does not seem to handle that kind of configuration very well. So at the moment, the way the Airflow Helm chart is done, it's either Airflow working, or Flower working, but not both...

TL;DR

  * Flower should log connection issues to the broker
  * it should also 'look' at TLS connections and at least warn us if it does not handle it properly
  * ... or even better, know how to automatically switch to TLS-enabled (with the system's certificates) if the remote endpoint asks for it

**Comment by mher** [comment]
Please try the latest master version. I've pushed a fix for this issue.

**Comment by northeastsquare** [comment]
@mher  I clone down latest master, but no use

**Comment by doizuc** [comment]
On macOS, flower was hanging for ever. The solution was to enable rabbitmq_management.

     sudo rabbitmq-plugins enable rabbitmq_management  

I had the following versions:

    amqp==5.1.0
    celery== 5.2.3
    flower==1.0.0
    kombu==5.2.4
    tornado==6.1

   RabbitMQ version: 3.9.13

I found the error through Ubuntu as flower was not hanging there but had the error below:

    [E 220313 19:31:31 broker:79] RabbitMQ management API call failed: [Errno 104] Connection reset by peer

**Comment by opyate** [comment]
In 2025, [this suggestion](https://github.com/mher/flower/issues/639#issuecomment-586535205) worked for me 🤔 

<details>

<summary>my env</summary>

pip freeze on flower docker container:

```
amqp==5.1.1
billiard==4.1.0
celery==5.3.1
click==8.1.6
click-didyoumean==0.3.0
click-plugins==1.1.1
click-repl==0.3.0
flower==2.0.1
humanize==4.7.0
kombu==5.3.1
prometheus-client==0.17.1
prompt-toolkit==3.0.39
python-dateutil==2.8.2
pytz==2023.3
redis==4.6.0
six==1.16.0
tornado==6.3.3
tzdata==2023.3
vine==5.0.0
wcwidth==0.2.6
```

docker compose config:

```yml
services:
  rabbitmq:
    image: rabbitmq:4.1-management
    ports:
      - "5672:5672"
      - "15672:15672"
    environment:
      RABBITMQ_DEFAULT_USER: "admin"
      RABBITMQ_DEFAULT_PASS: "admin"
    healthcheck:
      test: [ "CMD", "rabbitmq-diagnostics", "check_port_connectivity" ]
      interval: 5s
      timeout: 5s
      retries: 5
      start_period: 10s

  celery_worker:
    command: celery -A my_workers.celery_app worker -l INFO
    build:
      context: .
    depends_on:
      rabbitmq:
        condition: service_healthy
      redis:
        condition: service_healthy
    environment:
      CELERY_BROKER_URL: "amqp://admin:admin@rabbitmq:5672//"
      CELERY_RESULT_BACKEND: "redis://redis:6379/0"
    develop:
      # https://docs.docker.com/compose/file-watch/#compose-watch-versus-bind-mounts
      watch:
        - action: sync+restart
          path: ./src
          target: /app
        - action: rebuild
          path: ./uv.lock

  redis:
    image: redis:7.4
    ports:
      - "6379:6379"
    healthcheck:
      test: [ "CMD", "redis-cli", "ping" ]
      interval: 5s
      timeout: 5s
      retries: 5
      start_period: 3s

  redisinsight:
    image: redis/redisinsight:latest
    environment:
      RI_REDIS_HOST0: redis
    ports:
      - "5540:5540"

  flower:
    image: mher/flower:2.0.1
    environment:
      CELERY_BROKER_URL: "amqp://admin:admin@rabbitmq:5672//"
      CELERY_RESULT_BACKEND: "redis://redis:6379/0"
      FLOWER_BROKER_API: "http://admin:admin@rabbitmq:15672/api/"
      FLOWER_ADDRESS: "0.0.0.0"
      FLOWER_DEBUG: false
      FLOWER_PERSISTENT: true
      FLOWER_STATE_SAVE_INTERVAL: 10000
    ports:
      - "5556:5555" # <-- This line fixed it, and I can see the dashboard on localhost:5556
    depends_on:
      rabbitmq:
        condition: service_healthy
      redis:
        condition: service_healthy

```

</details>
